### PR TITLE
Visualstyle subclassing enabled, fixes #144

### LINKED ISSUE #144: Make visual style more easily configurable
Maybe I'm missing something obvious here, but I can't figure out how to change minor things in the DefaultVisualStyle.

In my current project, I simply want to change the alert's background color, and had hoped to be able to do

    class MyVisualStyle: DefaultVisualStyle {
        var backgroundColor: UIColor? {
            return UIColor.whiteColor()
        }
    }

but that doesn't work. Implementing the full VisualStyle protocol doesn't seem to work either, since AlertController.visualStyle is of type `DefaultVisualStyle`.

### PR DESCRIPTION
**Pull request by gereons**

### CODE REVIEW

**Comment by sberrevoets** [comment]
It's not immediately clear to me how this solves the problem, can you explain that?

**Comment by gereons** [comment]
Moving all implementation methods from the VisualStyle extension to the DefaultVisualStyle class makes them overridable, so my example from #144 now looks like
```
class MyVisualStyle: DefaultVisualStyle {
    override var backgroundColor: UIColor? { // now needs override keyword
        return UIColor.whiteColor()
    }
}
```
and works. This is the only way I found that allows me to override the methods from the extension. I also think it makes the code cleaner, because now there's just one procotol and one implementation of everything in it.

Changing the type of `AlertController.visualStyle` to be `VisualStyle` instead of `DefaultVisualStyle`  re-enables the alternative method of providing a visual style by proving a complete implementation of the protocol, as described in the README.

**Comment by sberrevoets** [comment]
I see. I'm gonna look into this some more and will get back to you. Thanks for your contribution!

**Comment by pixelmatrix** [comment]
I've been having issues with this as well. +1 for this fix!

**Comment by sberrevoets** [comment]
I dove into this a little bit and it does make sense why this isn't working. The problem with this change, however, is that making the `AlertController.visualStyle` of type `VisualStyle`, the interop with Objective-C is lost since protocols don't get translated to Objective-C land.

This is a tough one, because on one hand I think it's nice that the library works (or is supposed to work) the same in Objective-C and Swift, but in this case it's at the cost of not being able to use a protocol where I'd like to use one.

**Comment by gereons** [comment]
I'm not sure I understand how ObjC interop plays a role here. ObjC classes cannot subclass Swift classes, so overriding stuff from `DefaultVisualStyle` doesn't work. Likewise, implementing `VisualStyle` doesn't work since it's not exposed, as far as I can tell.

**Comment by sberrevoets** [comment]
That's a very good point, I completely forgot subclassing Swift classes from Objective-C isn't possible. In that case, I see a few options:

- Make the visual style a class written in Swift with settable properties
- Make the visual style a class written in Objective-C (using settable properties or subclassing)
- Expose the `alertStyle` as an `@objc protocol`.

 On the Objective-C side you would then have to implement *all* the required properties even if you only want to change 1 of them. On the Swift side default implementations wouldn't be possible anymore, so it appears that doing that would be some sort of "worst of both worlds" situation.

- Drop support for visual styling in Objective-C

I really like none of these options, but I guess I would prefer the first one over the others. Do you see other options?

**Comment by pixelmatrix** [comment]
Just to chime in here, I think the first is the best option. The only reason this impacts me at all is that it's totally impossible to change the background color now.

**Comment by gereons** [comment]
Option 1 would be best, I agree.

**Comment by sberrevoets** [comment]
Sounds good, let's go with that then.

@gereons, could you update your PR to reflect these changes?

**Comment by gereons** [comment]
Already on it. 

I'm unsure how you want to deal with the protocol's prescription of "all properties are read-only" if everything is converted to a `public var` - should I get rid of it entirely?

**Comment by sberrevoets** [comment]
Yes, I think that would be best. Now that changing properties is possible (without having to pass in a completely new visual style object), we should probably also document that changing properties on a visual style _after_ the alert was presented has no effect.

**Comment by gereons** [comment]
One further change I'd like to propose is renaming `DefaultVisualStyle` to `AlertVisualStyle` (prefixed, of course, for ObjC). With the new usage pattern, the `Default` parts seems inappropriate to me.

**Comment by sberrevoets** [comment]
Yep, seems like a good idea. Since these are all breaking changes let's get it all over with now. Still not quite sure at what point to release this because of necessary major version increase.

**Comment by gereons** [comment]
I don't think these changes will affect too many users, since customizing the layout wasn't really possible in Objective-C and pretty hard to do in Swift, and there haven't really been many complaints about it, as far as I can tell.

File: Source/AlertController.swift [hunk 1]
```diff
@@ -108,7 +108,7 @@ public class AlertController: UIViewController {
     public var shouldDismissHandler: (AlertAction? -> Bool)?
 
     /// The visual style that applies to the alert or action sheet.
-    public lazy var visualStyle: DefaultVisualStyle = DefaultVisualStyle(alertStyle: self.preferredStyle)
+    public lazy var visualStyle: AlertVisualStyle = AlertVisualStyle(alertStyle: self.preferredStyle)
```
Comment (sberrevoets) [inline Source/AlertController.swift]: The explicit type here isn't necessary anymore I believe, right?

**Comment by sberrevoets** [comment]
I'd like to at least explore the possibility of marking the current `visualStyle` deprecated so it doesn't need a major release. That may prove to be a larger maintenance burden than it's worth, but otherwise this would be the 3rd major release in 4 months which isn't great.

File: Source/AlertController.swift (hunk 1 above)
Comment (gereons) [inline Source/AlertController.swift]: I thought so too, but it is. Removing it leads to `Value of type 'NSObject -> () -> AlertController' has no member 'preferredStyle'`

**Comment by sberrevoets** [comment]
Found a way to keep the old class names around and deprecate them so this is only a minor update. Thanks for your effort and sorry for the delay!

**Pull request merged**